janusgraph-core/src/main/java/org/janusgraph/graphdb/util/SubqueryIterator.java

```diff
@@ -17,81 +17,236 @@
+    }
+
+    private void executeQueries(List<JointIndexQuery.Subquery> queries, final int idx, final int subLimit,
+                                int[] offsets, Map<Object, List<Integer>> subResultToQueryMap,
+                                IndexSerializer indexSerializer, BackendTransaction tx) {
+        JointIndexQuery.Subquery subQuery = queries.get(idx).updateLimit(subLimit);
+        final List<Object> subQueryCache = indexCache.getIfPresent(subQuery);
+        if (subQueryCache != null) {
+            assert subQueryCache.size() >= offsets[idx];
+            for (int j = offsets[idx]; j < subQueryCache.size(); j++) {
+                subResultToQueryMap.computeIfAbsent(subQueryCache.get(j), k -> new ArrayList<>()).add(idx);
+            }
+            offsets[idx] = subQueryCache.size();
+        } else {
+            QueryProfiler profiler = subQuery.getProfiler();
+            QueryProfiler.startProfile(profiler, subQuery);
+            // TODO: leverage the scrolling capability of indexing backends ratherthan throw away old results
+            indexSerializer.query(subQuery, tx).skip(offsets[idx]).forEachOrdered(result -> {
+                offsets[idx]++;
+                subResultToQueryMap.computeIfAbsent(result, k -> new ArrayList<>()).add(idx);
+            });
             profiler.stopTimer();
+            profiler.setResultSize(offsets[idx]);
+            // TODO: should we put it into cache?
+        }
+
+    }
+
+    private void intersectResults(Map<Object, List<Integer>> subResultToQueryMap,
+                                  List<Object> results, final int limit, final int numberOfQueries) {
+        for (Iterator<Map.Entry<Object, List<Integer>>> it = subResultToQueryMap.entrySet().iterator();
+             it.hasNext(); ) {
+            Map.Entry<Object, List<Integer>> entry = it.next();
+            if (entry.getValue().size() == numberOfQueries) {
+                // this particular result satisfies every index query
+                results.add(entry.getKey());
+                if (results.size() >= limit) break;
+                it.remove();
+            }
+        }
+    }
+
+    /**
+     * Compute score for each query. Lower score means the query is more selective and more likely to
+     * be the bottleneck. Exhausted queries always have 0 score. Unless more factors are taken into consideration,
+     * at the moment it does not make sense to compare queries if we only have two.
+     */
+    private double[] computeScores(final int resultsExhaustedCount, boolean[] resultsExhausted,
+                                   int[] offsets, final int numberOfQueries, final int numberOfResults,
+                                   final int limit, Map<Object, List<Integer>> subResultToQueryMap) {
+        double[] scores = new double[numberOfQueries];
+        // TODO: take more factors into consideration, e.g. query latency
```

Comment:
If we can find a way to let users define score/priority for indices, we can solve https://github.com/JanusGraph/janusgraph/issues/270 here